transport/endpoint.go

```diff
@@ -0,0 +1,108 @@
+package transport
+
+import (
+	"context"
+	"net/http"
+	"github.com/playnet-public/apiexperiments/problems"
+	"github.com/playnet-public/apiexperiments/encode"
+	"go.uber.org/zap"
+)
+
+//HandlerFunc of the main service functions
+type HandlerFunc func(ctx context.Context, r *http.Request) (interface{}, error)
+
+//MiddlewareFunc of middlewares surrounding the main service
+type MiddlewareFunc func(ctx context.Context) error
+
+//Endpoint of an service
+type Endpoint interface {
+	//WithBefore places a middleware before the handler.
+	//WithBefore can be called multiple as every middleware is appended one after another
+	WithBefore(mw MiddlewareFunc) Endpoint
```

Comment:
Not a fan of the naming here.

Reply:
Changed to `PutMiddlewareBefore` and `PutMiddlewareAfter`